final_task/rss_reader/rss_reader.py

```diff
@@ -0,0 +1,239 @@
+                    links=[f"[{num}]: {link['href']}" for num, link in enumerate(entry['links'], 1)])
+            )
+    except KeyError as value:
+        logging.error(f"One of the entries does not have {value} key")
+        logging.info('Application ended')
+        sys.exit()
+
+    return articles
+
+
+def print_rss_json(data: list) -> None:
+    """Prints result as JSON in stdout"""
+    print(json.dumps(data, cls=EnhancedJSONEncoder))
+
+
+def print_article(data: SingleArticle) -> None:
+    """Prints a single article in stdout"""
+    print("-" * 120)
+    print(data)
+    print("-" * 120)
+
+
+def print_rss_articles(rss_articles: list) -> None:
+    """Prints source feed and articles"""
+    for article in rss_articles:
+        print_article(article)
+
+
+def check_if_verbose(args: argparse.Namespace) -> None:
+    """Checks if verbose is on and prints logs to stdout"""
+    if args.verbose:
+        with open(LOG_FILE_NAME, 'r') as fin:
+            print(fin.read(), end='')
+        logging.info('Log was printed in stdout')
+
+
+def main(args: argparse.Namespace) -> None:
+    """The main entry point of the application"""
+    rss_url = args.source
+    limit = args.limit
+
+    parsed_rss = parse_rss(rss_url)
+    logging.info('RSS URL parsed')
+
+    articles_list = get_articles(parsed_rss, limit)
+    logging.info('Articles list created')
+
+    try:
+        if args.json:
+            print_rss_json(articles_list)
+            logging.info('Articles was printed as json')
+        else:
+            print_rss_articles(articles_list)
+            logging.info(f'{limit} articles was printed')
+    except IndexError:
+        logging.error('There is no articles by this URL')
+
+    check_if_verbose(args)
+
+    logging.info('Application ended')
+
+
+def check_the_connection() -> str:
+    """Checks the connection"""
+    info = ''
+    try:
+        urllib.request.urlopen(sys.argv[1])
+    except urllib.request.HTTPError as e:
+        info = f'{e.code}: {e.reason}'
+        return info
+    except urllib.request.URLError as e:
+        info = f'{e.reason}'
+        return info
+    else:
+        return info
+
+
+def check_if_link_is_correct() -> bool:
+    """Checks if link is valid"""
+    try:
+        result = urllib.request.urlparse(sys.argv[1])
+        return all([result.scheme, result.netloc, "." in result.netloc, len(result.netloc) > 2])
+    except ValueError:
+        return False
+
+
+def check_url_and_connection() -> None:
+    """Checks if url is valid and connection successful"""
+    if not check_if_help_or_version_in_arguments():
+        if check_if_link_is_correct():
+            logging.info('Valid url')
+            connection_info = check_the_connection()
+            if connection_info:
+                print('Connection failed')
+                logging.error(f"Connection failed: {connection_info}")
+                logging.info('Application ended')
+                sys.exit()
```

Comment:
Here and below: It is prefereble to have only one exit point from the application. Instead of `sys.exit()` here it is better to raise some exception(probably your custom exception) and handle it in the main function.